### PR TITLE
Add `mres` and `nres` Singular Calls for Free resolutions

### PR DESCRIPTION
**Pull request by RafaelDavidMohr**
Adds corresponding keywords to `free_resolution` and adjusts docu accordingly.

### CODE REVIEW

File: src/Modules/UngradedModules.jl [hunk 1]
```diff
@@ -6361,6 +6362,10 @@ function free_resolution(M::SubquoModule{<:MPolyRingElem};
     res = Singular.fres(singular_kernel_entry, length, "complete")
   elseif algorithm == :sres
     res = Singular.fres(singular_kernel_entry, length)
+  elseif algorithm == :mres
```
Comment (fingolfin) [inline src/Modules/UngradedModules.jl]: Don't these also require a recent enough Singular.jl? If so then Project.toml should be adjusted, too

Comment (RafaelDavidMohr) [reply src/Modules/UngradedModules.jl]: Sure! I am not sure I am doing this correctly in https://github.com/oscar-system/Oscar.jl/pull/2500/commits/e72cb47fb35fe8d53dbb9deee7dec1d77582bf52, do I just by hand bump the compat entry in Project.toml? Never had to fiddle around with this before.

File: src/Modules/UngradedModules.jl [hunk 2]
```diff
@@ -6361,6 +6362,10 @@ function free_resolution(M::SubquoModule{<:MPolyRingElem};
     res = Singular.fres(singular_kernel_entry, length, "complete")
   elseif algorithm == :sres
     res = Singular.fres(singular_kernel_entry, length)
```
Comment (wdecker) [inline src/Modules/UngradedModules.jl]: There seems to be an old bug: you want sres but ask for fres!

**Comment by wdecker** [comment]
Once this is merged, I will extend the docu and explain what the difference between the different methods is.

File: src/Modules/UngradedModules.jl (hunk 2 above)
Comment (RafaelDavidMohr) [reply src/Modules/UngradedModules.jl]: Fixed in https://github.com/oscar-system/Oscar.jl/pull/2500/commits/d6abd9fff03290d7a8cf394abccbfe60a73ae88d

File: Project.toml [hunk 3]
```diff
@@ -35,7 +35,7 @@ Polymake = "0.10.0"
 Preferences = "1"
 RandomExtensions = "0.4.3"
 RecipesBase = "1.2.1"
-Singular = "0.18.2"
+Singular = "0.18.6"
```
Comment (fingolfin) [inline Project.toml]: This looks right (I didn't check the versions, but I trust you did).

Thanks!

**Comment by fingolfin** [comment]
This PR has no test for the new code, so naturally it will pass, even if actually calling the code was broken.

So either a test should be added in this PR (could be a doctest or could be a regular test); or else I'd hope that @wdecker would add one when he extends the documentation for this?

**Comment by RafaelDavidMohr** [comment]
A word on the now failing tests: I could be wrong, but to me it seems that the issue is the following:

Oscar starts computing a free resolution by computing the input module as a cokernel and then passes that cokernel to Singular to let it compute a free resolution for that. `Singular.sres` returns this free resolution but with the generators of this cokernel potentially permuted. Now the presentation of the input module and the Singular resolution don't "patch together" correctly and thus give a wrong free resolution in Oscar. 

My question is: Should we try to handle this bug in this PR or a seperate one?

**Review by wdecker — COMMENTED** [review]
There are several problems here: To begin with, fres and sres require as input a gröbner basis. This is missing so far. There is another problem on which @jankoboehm is already working. 

For the time being, I suggest to replace sres by fres in the test. If this makes the test pass, let us merge the PR. More tests and docu will be added as soon as all problems are fixed. @RafaelDavidMohr please open an issue which reminds us to do the above.

**Comment by RafaelDavidMohr** [comment]
Unfortunately the same problem seems to be happening with `mres`: Singular changes the cokernel representation of the module and then the free resolution Singular gives doesn't glue correctly with what is computed on the Oscar side.

So `mres` would require similar changes to the free resolution code that `sres` needs.

**Comment by wdecker** [comment]
The situation for `mres` is different and more subtle as `mres` computes a new presentation. In particular, this requires that we return an isomorphism between the given module and the new module. That means that we have to introduce some book-keeping in Singular and add the result of this to the output of `mres`. So let me suggest to make this PR a draft until all these prblems have been addressed.

**Comment by RafaelDavidMohr** [comment]
> The situation for `mres` is different and more subtle as `mres` computes a new presentation. In particular, this requires that we return an isomorphism between the given module and the new module. That means that we have to introduce some book-keeping in Singular and add the result of this to the output of `mres`. So let me suggest to make this PR a draft until all these prblems have been addressed.

Ok! In a sence, `sres` also computes a new representation (but just by shuffling around generators). 

Let me know if I can help with fixing this.

**Comment by fingolfin** [comment]
We think there are some changes in the latest Singular.jl that may help, but couldn't find them just now; but @hannes14 will check

**Comment by RafaelDavidMohr** [comment]
I added `mres`, the bug that I originally described (with Singular changing the presentation of some module) looks to be actually easily fixed by some global changes to the `free_resolution` function, we just have to also take the first map that Singular gives us.

Let me know if I missed something.

**Comment by ederc** [comment]
Looks like a valid solution to me.

File: test/Modules/UngradedModules.jl [hunk 4]
```diff
@@ -171,16 +171,20 @@ end
 	B = R[x^2; x*y; y^2; z^4]
 	M = SubquoModule(A, B)
 	free_res = free_resolution(M, length=1)
-    @test is_complete(free_res) == false
-    free_res[2]
-    @test length(free_res.C.maps) == 4
+  @test is_complete(free_res) == false
+  free_res[2]
+  @test length(free_res.C.maps) == 4
```
Comment (lgoettgens) [inline test/Modules/UngradedModules.jl:176]: This part seems to be indented weirdly (when viewing it on github). Maybe there is a mixup of tabs and spaces here?

Comment (RafaelDavidMohr) [reply test/Modules/UngradedModules.jl:176]: There were some 4 space indentations in this file which I removed, everything looks good in my local files.

Comment (lgoettgens) [reply test/Modules/UngradedModules.jl:176]: Let's just leave this as is and let https://github.com/oscar-system/Oscar.jl/pull/2785 handle the tab issue.

**Comment by fingolfin** [comment]
If the geometry would like to merge this, fine by me, but someone needs to explicitly approve it

**Comment by wdecker** [comment]
We are still discussing it. Will be merged tomorrow.

**Comment by RafaelDavidMohr** [comment]
Here is a bit more explanation on why things should work now. Originally I diagnosed the problem wrong.

Previously, for a given module `M`, `free_resolution(M)` would perform the following computations:

We start by computing a presentation of `M`
```
0 <-- M <-- F <-- G
```
then we let `C` be the cokernel of this representation (i.e. the image of `G` in `F`) and let Singular compute a free resolution of this cokernel:
```
0 <-- F/C <-- F0 <-- F1 <-- F2 <-- ...
```
and now the two chain complexes were patched together as follows:
```
0 <-- M <-- F <-- G <-- F2 <-- ...
```
`F` and `F0` are always the same, but the problematic part is `F <-- G <-- F2` because the image of  `G` in `F` (`==F0`) might have its generators sorted differently or altogether changed  compared to the image of `F1` in `F0`. Then the homology is nonzero at `G`. This is now fixed by instead returning
```
0 <-- M <-- F <-- F1 <-- F2 <-- ...
```
This is now always a correct free resolution (because we basically just take the one Singular gives us, `F = F0` is always guaranteed). I hope this helps.

**Comment by wdecker** [comment]
@RafaelDavidMohr Thx for doing this.

I have a small remark: The methods

   `free_resolution(I::MPolyIdeal)`

   `free_resolution(Q::MPolyQuoRing)`

and also the "dummy" method

   `free_resolution(F::FreeMod)`

should be adjusted to the syntax we have now for `free_resolution`.

**Comment by jankoboehm** [comment]
As I understand it, we switch the presentation matrix originally computed by OSCAR to the (in case of mres, not fres) possibly modified one returned by Singular. Are we sure that this is the same module and not just an isomorphic one (then we have to keep track of the isomorphism), what if we resolve a submodule. We cannot change the generators, since the representation of the elements is referencing them. It also looks to me as if the grading is removed? Best have a quick talk in person once possible.

**Comment by RafaelDavidMohr** [comment]
> As I understand it, we switch the presentation matrix originally computed by OSCAR to the (in case of mres, not fres) possibly modified one returned by Singular. Are we sure that this is the same module and not just an isomorphic one (then we have to keep track of the isomorphism), what if we resolve a submodule. We cannot change the generators, since the representation of the elements is referencing them. It also looks to me as if the grading is removed? Best have a quick talk in person once possible.

Yes, in person (Saarbrücken for instance) is probably easier. I will however have a look at the grading issue and at the things @wdecker mentioned.

Edit: Actually, @jankoboehm, I am not sure what you mean by "the grading is removed". I tried a random example and I'm getting a graded free resolution. Could you specify?

**Comment by RafaelDavidMohr** [comment]
> @RafaelDavidMohr Thx for doing this.
> 
> I have a small remark: The methods
> 
> `free_resolution(I::MPolyIdeal)`
> 
> `free_resolution(Q::MPolyQuoRing)`
> 
> and also the "dummy" method
> 
> `free_resolution(F::FreeMod)`
> 
> should be adjusted to the syntax we have now for `free_resolution`.

See https://github.com/oscar-system/Oscar.jl/pull/2500/commits/347349cfef508bef64bc977876118e02bf0efccc. I adjusted the docu accordingly, the docu for `free_resolution(F::FreeMod)` is not ideal, let me know if I should change it.

**Comment by wdecker** [comment]
This is good for now. However, for the `mres` part, we also need to minimize the presentation we work with. This requires some work in Singular first. We will fix this asap.

**Review by wdecker — APPROVED** [review]

**Pull request merged**